### ISSUE TITLE
Local probability versus number of samples

### ISSUE THREAD

**Issue by VincentBonnivard** [marcotcr/lime#149]
Hello,

I'm using LIME for interpreting a classifier. The model is a XGBoost gradient boosting, trained on 5387 instances, using 30 features.
I'm looking at the impact on the local probability of varying the number of samples used for the local interpretation. 

On the Figure below, you can see for a typical instance the local probability (in blue) versus the true probability (in green), when varying the number of samples from 100 to 5000 (in details, we actually generated 50 explanations for each number of samples, and looked at the mean local probability and its standard deviation, in orange).

![lime_distrib_proba](https://user-images.githubusercontent.com/35725953/35269319-f199aaa0-002b-11e8-98cf-f0cb9ab7ed43.png)

As you can see, the larger is the number of samples, the larger is the deviation to the true probability.

However, we also looked at the explanations themselves, and the larger the number of the samples, the more "robust" they appear (i.e. less variance in terms of order of features appearing in the explanation).

So, should the deviation between local and true probabilities be used to measure the robustness of the local interpretation ? Or can the interpretation be only better if we increase n_samples ?

Thanks!

**Comment by echo66** [comment]
Did you decrease the kernel_width?

**Comment by VincentBonnivard** [comment]
No, we let the default value (None). Here is the code we use:

`explainer = lime.lime_tabular.LimeTabularExplainer(training_data=X_train,
                                                   feature_names=important_features,
                                                   class_names=['good', 'bad'],
                                                   discretize_continuous=True,
                                                   discretizer='decile',
                                                   verbose=True,
                                                   mode='classification')
`

`exp = explainer.explain_instance(X,
                                 xgb_mod.predict_proba,
                                 num_features=30,
                                 num_samples=50)`

**Comment by echo66** [comment]
If you reduce the kernel width, you will make decrease the divergence between the local prediction and original prediction **but** you will probably decrease the R² score of the explanation.

**Comment by VincentBonnivard** [comment]
Thanks for your answer. Actually, the question is: how is the sampling done ? I was expecting it to be uniform, but it looks like the more points you generate, the further away from the observation they're likely to be

**Comment by VincentBonnivard** [comment]
I looked at the way the sampling is done. If I understood correctly:
   - Each feature is standard scaled using the training data,
   - n_samples are generated on a Normal distribution with mean 0, std 1,
   - the n_samples are then put back into original space.

I'm a bit confused about this approach, because the samples are not generated close to the instance under investigation. Of course they are then weighted, but it looks not optimal to me if we want for instance to explain an outlier (which is likely to be of high interest): most of the generated samples will be far from the observation.

Is there a reason why not to sample in the vicinity of the instance for which we need an explanation ?

Thanks!

**Comment by VincentBonnivard** [comment]
No problem ! We have implemented something that uses 50 interpretations for each instance, with 3000 samples, with kernel = 0.1*sqrt(n_var). The local probability is in good agreement with the true one, and the interpretations appear to be consistent.

So it's not urgent, but we'd be happy to look at your blog post!
Thanks

**Comment by narendiran1996** [comment]
I looked at the way the sampling is done. If I understood correctly:

Each feature is standard scaled using the training data,
n_samples are generated on a Normal distribution with mean 0, std 1,
the n_samples are then put back into original space.
I'm a bit confused about this approach, because the samples are not generated close to the instance under investigation. Of course they are then weighted, but it looks not optimal to me if we want for instance to explain an outlier (which is likely to be of high interest): most of the generated samples will be far from the observation.

Is there a reason why not to sample in the vicinity of the instance for which we need an explanation ?

@echo66 
Hey Vicent! Sorry for the delay! 😕

I will do an in-depth blog post about this subject soon. How urgent is your need?


Any update on this??

**Comment by remindchobits** [comment]
Hi, all

I also have the problem, does R2 score really works? as number of samples increases, the (weighted) averaged L1 distance between the predictions and labels get closer, but R2 score decreases.

Does anyone find the solution?
